packages/doenetml-worker-rust/lib-doenetml-core/src/core/state_var_interfaces/text_state_var_interfaces.rs

```diff
@@ -0,0 +1,197 @@
+        }
+
+        dep_instructs
+    }
+
+    fn save_dependencies_for_value_calculation(&mut self, dependencies: &Vec<Vec<Dependency>>) {
+        let num_dependencies = dependencies.iter().fold(0, |a, c| a + c.len());
+
+        let mut string_vals = Vec::with_capacity(num_dependencies);
+
+        for instruction in dependencies.iter() {
+            for Dependency {
+                value: dep_value, ..
+            } in instruction.iter()
+            {
+                if let StateVarReadOnlyView::String(dep_string_value) = dep_value {
+                    string_vals.push(dep_string_value.create_new_read_only_view())
+                } else {
+                    panic!("Got a non-string value for a dependency for a GeneralStringStateVarInterface");
+                }
+            }
+        }
+
+        self.string_dependency_values = string_vals;
+
+        if num_dependencies == 1 {
+            if let DependencySource::Essential { .. } = dependencies[0][0].source {
+                self.from_single_essential = true;
+            }
+        }
+    }
+
+    fn calculate_state_var_from_dependencies_and_mark_fresh(
+        &self,
+        state_var: &StateVarMutableViewTyped<String>,
+    ) {
+        if self.from_single_essential {
+            // if we are basing it on a single essential variable,
+            // then we propagate used_default as well as the value.
+            state_var.set_value_and_used_default(
+                self.string_dependency_values[0].get_fresh_value().clone(),
+                self.string_dependency_values[0].get_used_default(),
+            );
+        } else {
+            // TODO: can we implement this without cloning the inner value?
+            let value: String = self
+                .string_dependency_values
+                .iter()
+                .map(|v| v.get_fresh_value().clone())
+                .collect();
+
+            state_var.set_value(value);
+        }
+    }
+
+    fn request_dependencies_to_update_value(
+        &self,
+        state_var: &StateVarReadOnlyViewTyped<String>,
+        _is_direct_change_from_renderer: bool,
+    ) -> Result<Vec<DependencyValueUpdateRequest>, RequestDependencyUpdateError> {
+        if self.string_dependency_values.len() != 1 {
+            // TODO: implement for no dependencies where saves to essential value?
+            Err(RequestDependencyUpdateError::CouldNotUpdate)
+        } else {
+            let desired_value = state_var.get_requested_value();
+
+            self.string_dependency_values[0].request_change_value_to(desired_value.clone());
+
+            Ok(vec![DependencyValueUpdateRequest {
+                instruction_idx: 0,
+                dependency_idx: 0,
+            }])
+        }
+    }
+}
+
+/// A simplified version of GeneralStringStateVarInterface
+/// that is based on a single dependency.
+/// Requires a `dependency_instruction_hint`.
+///
+/// Panics if `create_dependency_from_extend_source` is true,
```

Comment:
I think we should shift to returning errors whenever we can.

Reply:
This is a programming error if someone made a mistake, so I'd want it to blow under these conditions. The right behavior is that the whole document should disappear and be replaced by an ugly error message. One option is that the worker quits and send a message to the renderer. We can certainly come up with better way to do this with propagating errors. I could create an issue with the task of determining the best approach.

Reply:
An issue sounds good. A common pattern is for there to be a `dev mode` where errors are fatal, but then have a `production mode` where any recoverable errors are delt with. It would be nice for your problem not to dissapear in the middle of a midterm because of one of your mistakes...